### PR TITLE
Simulated network environment

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

**Comment by nfrisby** [comment]
(I removed Pawel and myself from the reviewers list. I'm guessing Marcin opened the PR before rebasing and so the CODEOWNERS added us when the PR originally touched some Consensus files. It doesn't touch any such files now.)

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+--
+-- AttenuatedChannel
+--
+
+
+-- | An AttenuatedChannel supports:
+--
+-- - attenuation applied after reading a message from 'QueueChannel';
+-- - two-way close handshake with 120s timeout.  Read side is closed as soon as
+--   an internal 'MsgClose' is received, write side has to be closed explicietly.
```
Comment (bolt12) [inline network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:122]: Small typo: `explicietly` -> `explicitly`

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+                             , acWrite = acWrite writeCounterVar
+                             , acClose
+                             }
+  where
+    acRead :: m BL.ByteString
+    acRead = do
+      msg <- atomically $ readQueueChannelSTM qc
+      t <- getMonotonicTime
+      case msg of
+        -- match the 'Bearer.Snocket' behavour and throw 'MuxError'
```
Comment (bolt12) [inline network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:181]: Typo: `behavour` -> `behavior`

Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs]: Fixed.

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs (hunk 1 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs]: Fixed

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+data SuccessOrFailure = Success | Failure
+
+type Size = Int64
+
+-- | Attenuation of a channel.
+--
+data Attenuation = Attenuation {
+    aReadAttenuation  :: Time -> Size -> ( DiffTime,
+                                           SuccessOrFailure ),
```
Comment (bolt12) [inline network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:150]:
```haskell
aReadAttenuation  :: Time -> Size -> ( DiffTime,
                                       SuccessOrFailure ),
```

->

I tend to prefer
```haskell
aReadAttenuation  :: Time -> Size -> ( DiffTime
                                     , SuccessOrFailure ),
```

File: ouroboros-network-framework/src/Ouroboros/Network/IOSim.hs [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+--
+data BearerInfo = BearerInfo
+    {
+      -- | How long it take to create a connection
+      biConnectionDelay      :: !DiffTime
+
+      -- | attenuation of inbound side of the bearer, i.e. attenuation used by
+      -- bearers that were 'accept'ed.
+    , biInboundAttenuation   ::  Time -> Size -> ( DiffTime,
+                                                   SuccessOrFailure )
```
Comment (bolt12) [inline ouroboros-network-framework/src/Ouroboros/Network/IOSim.hs:119]: Here as well

File: ouroboros-network-framework/src/Ouroboros/Network/IOSim.hs [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+
+      -- | attenuation of outbound side of the bearer, i.e. the attenuation used
+      -- by bearers that were created with 'connect' call.
+      --
+    , biOutboundAttenuation  ::  Time -> Size -> ( DiffTime,
+                                                   SuccessOrFailure )
```
Comment (bolt12) [inline ouroboros-network-framework/src/Ouroboros/Network/IOSim.hs:125]: As well

**Review by bolt12 — APPROVED** [review]
LGTM

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs (hunk 3 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:150]: ok, I'll switch.

**Review by karknu — COMMENTED** [review]
LGTM

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+--
+data CloseError = CloseTimeoutError
+  deriving (Show, Typeable)
+
+instance Exception CloseError
+
+
+data SuccessOrFailure = Success | Failure
+
+type Size = Int64
```
[the remark this answers is not in the document]
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:144]: I think we need to make a PR to `bytestring` to switch to `Word64`.

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+      sent <- writeQueueChannel qc (MsgBytes bs)
+      when (not sent) $
+        throwIO (resourceVanishedIOError "AttenuatedChannel.write")
+
+    acClose :: m ()
+    acClose = do
+      sent <- writeQueueChannel qc MsgClose
+      traceWith tr (AttenuatedChannelClosing sent)
+
+      -- TODO: switch to timeout once it's fixed.
```
[the remark this answers is not in the document]
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:220]: This is fixed in `master` in https://github.com/input-output-hk/ouroboros-network/commit/f460d2f34f92b734092bdeb211f41e3c9a18cd4d, but not yet in `p2p-master`.

**Pull request merged**